mass_mailing_custom_unsubscribe/tests/test_ui.py

```diff
@@ -0,0 +1,176 @@
+# Copyright 2016 Jairo Llopis <jairo.llopis@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+import mock
+from contextlib import contextmanager
+from odoo.tests.common import HttpCase
+from werkzeug import urls
+
+
+class UICase(HttpCase):
+    _tour_run = "odoo.__DEBUG__.services['web_tour.tour'].run('%s')"
+    _tour_ready = "odoo.__DEBUG__.services['web_tour.tour'].tours.%s.ready"
+
+    def extract_url(self, mail, *args, **kwargs):
+        url = mail._get_unsubscribe_url(self.email)
+        self.assertTrue(urls.url_parse(url).decode_query().get('token'))
+        self.assertTrue(url.startswith(self.domain))
+        self.url = url.replace(self.domain, "", 1)
+        return True
+
+    def setUp(self):
+        super(UICase, self).setUp()
+        self.email = "test.contact@example.com"
+        self.mail_postprocess_patch = mock.patch(
+            "odoo.addons.mass_mailing.models.mail_mail.MailMail."
+            "_postprocess_sent_message",
+            autospec=True,
+            side_effect=self.extract_url,
+        )
+        with self.tempenv() as env:
+            self.domain = env["ir.config_parameter"].get_param('web.base.url')
+            List = self.lists = env["mail.mass_mailing.list"]
+            for n in range(4):
+                self.lists += List.create({"name": "test list %d" % n})
+            self.contact = env["mail.mass_mailing.contact"].create({
+                "name": "test contact",
+                "email": self.email,
+                "list_ids": [(6, False, self.lists.ids)],
+            })
+            self.mailing = env["mail.mass_mailing"].create({
+                "name": "test mailing %d" % n,
+                "mailing_model_id": self.env.ref(
+                    'mass_mailing.model_mail_mass_mailing_list').id,
+                "contact_list_ids": [
+                    (6, 0, [self.lists[0].id, self.lists[3].id])
+                ],
+                "reply_to_mode": "thread",
+            })
+            self.mailing._onchange_model_and_list()
+            # HACK https://github.com/odoo/odoo/pull/14429
```

Comment:
I wonder if this is still needed. Maybe you could try to remove it.

Reply:
That PR was made for version odoo:9.0 but is still needed for v12: https://github.com/odoo/odoo/blob/20eca78cb9afd6288f5900d4833dc83cbe4fa811/addons/mass_mailing/models/mass_mailing.py#L628